(PE-22724) Add PE managed postgres support

cthorn42 opened the pull request:
This PR adds support for PE managed postgres support.
It detects for a new role: pe_postgres.
If that role is found, that node will be set as the puppetdb PE node.
Then there will be a check for how to configure the postgres node.
Either way SSL or with password, depending on options supplied.
The meep 2.0 settings are a bit of work in progress, future work
will probably be required to get them just right.

puppetlabs-jenkins commented:
Can one of the admins verify this patch?

cthorn42 commented:
I'm sorry, I forgot I had one last check to do on this. Please consider this do not merge.

cthorn42 commented:
Alright, turns out helps to push the correct updated branch. Should be good to review.

ericwilliamson commented on lib/beaker-answers/pe_conf.rb:
```diff
@@ -82,6 +82,28 @@ def _generate_1_0_data
 
+      if any_host_with_role?('pe_postgres')
+        pe_conf["#{ns}::database_host"] = the_host_with_role('pe_postgres').hostname
+        if hosts[0].options[:pe_postgresql_options]
```
are the pe_postgresql_options being set at the global level, or are you going to set them per host? If per host, is there a reason to have it on the host level instead of global?

cthorn42 replied:
Good question. When I'm supplying an options file, Beaker's global options is empty, the options are set at the host level.

cthorn42 replied:
Am I doing something wrong there? Seems odd to me.

ericwilliamson commented on lib/beaker-answers/pe_conf.rb:
```diff
@@ -82,6 +82,28 @@ def _generate_1_0_data
+        pe_conf["#{ns}::database_host"] = the_host_with_role('pe_postgres').hostname
+        if hosts[0].options[:pe_postgresql_options]
+          if hosts[0].options[:pe_postgresql_options][:security] == 'cert'
+            pe_conf["#{ns}::database_ssl"] = true
```
this logic is duplicated, consider extracting it out into it's own methods, such as `postgres_cert_answers` and `postgres_password_answers` that returns a hash you can merge into the pe_conf one.

ericwilliamson commented on lib/beaker-answers/pe_conf.rb:
```diff
@@ -82,6 +82,28 @@ def _generate_1_0_data
+            pe_conf["#{ns}::database_ssl"] = true
+            pe_conf["#{ns}::database_cert_auth"] = true
+          elsif hosts[0].options[:pe_postgresql_options][:security] == 'password'
+            pe_conf["#{ns}::activity_database_password"] = "PASSWORD"
```
since we know have pe_postgresql_options, should we allow the password to be configurable?

cthorn42 replied:
I don't have a real strong opinion on this. It could be added, but not sure on the reason for?

ericwilliamson commented on lib/beaker-answers/pe_conf.rb:
```diff
@@ -162,5 +201,21 @@ def the_host_with_role(role)
+    # Used for optional roles that may or may not be present
+    #
+    # @return True if there is atleast one host with the supplied role
+    def any_host_with_role?(role)
```
why did you need to copy this from beaker?

cthorn42 replied:
uhhh...I didn't know that was a method in beaker. Makes sense though! Originally I was using the_host_with_role method, which broke specs since that will raise an error if there is no role defined. I'll

cthorn42 replied:
I'm having issues trying to get beaker-answers to inherit from beaker. So for the moment I'm leaving this in while you review my other changes.

ericwilliamson commented on spec/beaker-answers/pe_conf_spec.rb:
```diff
@@ -21,6 +21,23 @@
       end
     end
 
+    context 'pe_postgres' do
```
a) these specs don't pass

```

Failures:

  1) BeakerAnswers::PeConf 1.0 schema pe_postgres generates a hash of monolithic configuration data
     Failure/Error: if hosts[0].options[:pe_postgresql_options]

     NoMethodError:
       undefined method `options' for #<StringifyHash:0x007fc732b09040>
     Shared Example Group: "pe.conf configuration" called from ./spec/beaker-answers/pe_conf_spec.rb:89
     # ./lib/beaker-answers/pe_conf.rb:87:in `_generate_1_0_data'
     # ./lib/beaker-answers/pe_conf.rb:57:in `configuration_hash'
     # ./spec/beaker-answers/pe_conf_spec.rb:35:in `block (4 levels) in <top (required)>'

  2) BeakerAnswers::PeConf 2.0 schema pe_postgres generates a hash of monolithic configuration data
     Failure/Error: if hosts[0].options[:pe_postgresql_options]

     NoMethodError:
       undefined method `options' for #<StringifyHash:0x007fc731b81be0>
     Shared Example Group: "pe.conf configuration" called from ./spec/beaker-answers/pe_conf_spec.rb:131
     # ./lib/beaker-answers/pe_conf.rb:147:in `_generate_2_0_data'
     # ./lib/beaker-answers/pe_conf.rb:59:in `configuration_hash'
     # ./spec/beaker-answers/pe_conf_spec.rb:35:in `block (4 levels) in <top (required)>'

Finished in 0.92143 seconds (files took 0.24155 seconds to load)
224 examples, 2 failures

Failed examples:

rspec './spec/beaker-answers/pe_conf_spec.rb[1:1:2:1]' # BeakerAnswers::PeConf 1.0 schema pe_postgres generates a hash of monolithic configuration data
rspec './spec/beaker-answers/pe_conf_spec.rb[1:2:2:1]' # BeakerAnswers::PeConf 2.0 schema pe_postgres generates a hash of monolithic configuration data
```


b) You're not testing any of the conditional logic you added - please add tests for those

cthorn42 replied:
doh.

cthorn42 commented:
@ericwilliamson Okay I fixed the spec tests, and added in another one. There is one with cert and one with password. I figure that is enough for checking the conditional logic? Or is more required?

cthorn42 commented:
@ericwilliamson when you get a chance can you give this PR another once over?

jpartlow commented on lib/beaker-answers/pe_conf.rb:
```diff
@@ -157,10 +182,40 @@ def the_host_with_role(role)
+
+    def postgres_cert_answers(pe_conf, meep_schema_version)
+      case meep_schema_version
+      when '1.0'
```
Since there is no difference in what is being set relative to meep schema, there's no need to duplicate it.

For now, when '1.0','2.0' should work for both these methods.

cthorn42 replied:
I'm attempting to make it easier to maintain if we ever change those settings in the future for 2.0. If you think that probably wont happen anytime soon I can simplify this.

jpartlow replied:
For now I'd remove the duplication; leave the case and it's simple enough to change in the future if needed.

cthorn42 commented on lib/beaker-answers/pe_conf.rb:
```diff
@@ -157,10 +182,40 @@ def the_host_with_role(role)
       end
 
-      if found_hosts.length == 0 or found_hosts.length > 1
+      if (found_hosts.length == 0 or found_hosts.length > 1) and (role != 'pe_postgres')
```
If there is no host with pe_postgres then an error is raised. 
Since pe_postgres is optional, we shouldn't raise that error.
Hurm, yes I see your point, if we have more then one pe_postgres we should raise a specific error. I'll work on that and change out the &&

jpartlow replied:
I see, so we have two ways of calling the_host_with_role(), in one case we consider it exceptional if we can't find a node, and in the other we don't.  Rather than specialize around this method understanding about pe_postgres, let's pass it an option to raise_exceptions.  For more idiomatic ruby, could default that to false, and add a the_host_with_role!(role) which calls the_host_with_role(role, true), and sub the_host_with_role!() wherever we are currently using it (except the times we are checking for pe_postgres).

cthorn42 commented on lib/beaker-answers/pe_conf.rb:
```diff
@@ -82,6 +82,19 @@ def _generate_1_0_data
 
+      if the_host_with_role('pe_postgres')
+        pe_conf["#{ns}::database_host"] = the_host_with_role('pe_postgres').hostname
+        if hosts[0][:pe_postgresql_options]
```
Oh, that is a mistake. I think I did that when working out the specs.

highb commented on spec/beaker-answers/pe_conf_spec.rb:
```diff
@@ -97,4 +178,103 @@
+    it 'returns postgres node in monolithic if pe_postgres is set as a role' do
+      mono_hosts[1]['roles'] = ['pe_postgres','agent']
+      peconf = BeakerAnswers::PeConf.new(mono_hosts, "1.0")
+      expect(peconf.send(:the_host_with_role, 'pe_postgres', raise_error=false)).to eq(mono_basic_hosts[1])
```
Should `raise_error` be passed in here? We're not expecting a duplicate role in this test, are we?

cthorn42 replied:
Good point, thanks.

highb commented on lib/beaker-answers/pe_conf.rb:
```diff
@@ -82,6 +82,19 @@ def _generate_1_0_data
+        pe_conf["#{ns}::database_host"] = the_host_with_role('pe_postgres', raise_error=false).hostname
+        if options[:pe_postgresql_options]
+          if options[:pe_postgresql_options][:security] == 'cert'
+            postgres_cert_answers(pe_conf, "1.0")
```
![singlequotes](https://user-images.githubusercontent.com/759848/33956486-89c2ed2c-dff3-11e7-8c96-81c57506590f.jpg)

highb replied:
This repo should really have a linter.

highb replied:
Nevermind, I'll add a linter and auto-fix all this.

highb commented:
I think all the previously requested changes are in so 👍
